tests/store/test_ftp_artifact_repo.py

```diff
@@ -0,0 +1,169 @@
+
+    artifacts = repo.list_artifacts(path=dir_name)
+
+    ftp_mock.nlst.assert_called_once_with(artifact_root_path + dir_name)
+    ftp_mock.size.assert_called_once_with(artifact_root_path + dir_name + '/' + file_path)
+
+    assert len(artifacts) == 2
+    assert artifacts[0].path == dir_name + '/' + file_path
+    assert artifacts[0].is_dir is False
+    assert artifacts[0].file_size == file_size
+    assert artifacts[1].path == dir_name + '/' + subdir_name
+    assert artifacts[1].is_dir is True
+    assert artifacts[1].file_size is None
+
+
+def test_log_artifact(ftp_mock, tmpdir):
+    repo = FTPArtifactRepository("ftp://test_ftp/some/path", ftp_mock)
+
+    d = tmpdir.mkdir("data")
+    f = d.join("test.txt")
+    f.write("hello world!")
+    fpath = d + '/test.txt'
+    fpath = fpath.strpath
+
+    repo.log_artifact(fpath)
+
+    ftp_mock.mkd.assert_called_once_with('/some/path')
+    ftp_mock.cwd.assert_called_once_with('/some/path')
+    ftp_mock.storbinary.assert_called_once()
+    assert ftp_mock.storbinary.call_args_list[0][0][0] == 'STOR test.txt'
+
+
+def test_log_artifacts(ftp_mock, tmpdir):
+    repo = FTPArtifactRepository("ftp://test_ftp/some/path", ftp_mock)
+
+    subd = tmpdir.mkdir("data").mkdir("subdir")
+    subd.join("a.txt").write("A")
+    subd.join("b.txt").write("B")
+    subd.join("c.txt").write("C")
+
+    repo.log_artifacts(subd.strpath)
+
+    ftp_mock.mkd.assert_any_call('/some/path/subdir')
+    ftp_mock.cwd.assert_any_call('/some/path/subdir')
+    assert ftp_mock.storbinary.call_count == 3
+    storbinary_call_args = sorted([ftp_mock.storbinary.call_args_list[i][0][0] for i in range(3)])
+    assert storbinary_call_args == ['STOR a.txt', 'STOR b.txt', 'STOR c.txt']
+
+
+def test_download_artifacts_single(ftp_mock):
+    repo = FTPArtifactRepository("ftp://test_ftp/some/path", ftp_mock)
+
+    ftp_mock.cwd = MagicMock(side_effect=ftplib.error_perm)
+
+    repo.download_artifacts("test.txt")
+
+    ftp_mock.retrbinary.assert_called_once()
+    assert ftp_mock.retrbinary.call_args_list[0][0][0] == 'RETR /some/path/test.txt'
+
+
+def test_download_artifacts(ftp_mock):
+    repo = FTPArtifactRepository("ftp://test_ftp/some/path", ftp_mock)
+
+    # mocked file structure
+    #  |- model
+    #     |- model.pb
+    #     |- variables
+    #        |- test.txt
+    dir_name = 'model'
+
+    # list artifacts at sub directory level
+    file_path = 'model.pb'
+    subdir_name = 'variables'
+    subfile_name = 'test.txt'
+    ftp_dir_name = '/some/path/' + dir_name
+
+    ftp_mock.cwd = MagicMock(side_effect=[None, None, ftplib.error_perm,
+                                          None, None, ftplib.error_perm])
+    ftp_mock.nlst = MagicMock(side_effect=[[file_path, subdir_name], [subfile_name]])
+
+    repo.download_artifacts("model")
+
+    cwd_call_args = sorted([ftp_mock.cwd.call_args_list[i][0][0] for i in range(6)])
+    expect_args = [ftp_dir_name, ftp_dir_name, ftp_dir_name + '/' + file_path,
```

Comment:
Nit: could use `os.path.join` here to avoid having to write out all the "/" characters

Reply:
fixed